File: IGListTest/IGListTest/ViewControllers/FeedViewController.m
//
//  FeedViewController.m
//  IGListTest
//
//  Created by liuruixuan on 2023/4/16.
//

#import "FeedViewController.h"
#import <IGListKit/IGListKit.h>
#import "JournalEntryLoader.h"
#import "Pathfinder.h"
#import "MarslinkModel.h"
#import "MessageSectionController.h"
#import "WeatherSectionController.h"
#import "JournalSectionController.h"

@interface FeedViewController () <IGListAdapterDataSource, PathfinderDelegate>

@property (nonatomic, strong) JournalEntryLoader *loader;

@property (nonatomic, strong) UICollectionView *collectionView;

@property (nonatomic, strong) IGListAdapter *adapter;

@property (nonatomic, strong) Pathfinder *pathfinder;

@property (nonatomic, strong) Weather *currentWeather;

@end

@implementation FeedViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    [self.loader loadLatest];
    [self.view addSubview:self.collectionView];
    self.adapter.collectionView = self.collectionView;
    self.adapter.dataSource = self;
    self.pathfinder.delegate = self;
    [self.pathfinder connect];
}

- (void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];
    self.collectionView.frame = self.view.bounds;
}

- (NSArray<id<IGListDiffable>> *)objectsForListAdapter:(IGListAdapter *)listAdapter {
    NSMutableArray *items = [[NSMutableArray alloc] init];
    [items addObject:self.currentWeather];
    [items addObjectsFromArray:self.loader.entries];
    [items addObjectsFromArray:self.pathfinder.messages];
    
    [items enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        if ([obj isKindOfClass:[Weather class]] || [obj isKindOfClass:[JournalEntry class]] || [obj isKindOfClass:[Message class]]) {
            
        }
    }];
    
    return items;
    
//    [items sortUsingComparator:^NSComparisonResult(id  _Nonnull obj1, id  _Nonnull obj2) {
//
//    }];
}

- (IGListSectionController *)listAdapter:(IGListAdapter *)listAdapter sectionControllerForObject:(id)object {
    if ([object isKindOfClass:[Message class]]) {
        return [[MessageSectionController alloc] init];
    } else if ([object isKindOfClass:[Weather class]]) {
        return [[WeatherSectionController alloc] init];
    } else {
        return [[JournalSectionController alloc] init];
    }
}

- (UIView *)emptyViewForListAdapter:(IGListAdapter *)listAdapter {
    return nil;
}

- (void)pathfinderDidUpdateMessagesWithPathfinder:(Pathfinder *)pathfinder {
    [self.adapter performUpdatesAnimated:YES completion:nil];
}

- (JournalEntryLoader *)loader {
    if (!_loader) {
        _loader = [[JournalEntryLoader alloc] init];
    }
    return _loader;
}

- (UICollectionView *)collectionView {
    if (!_collectionView) {
        UICollectionViewFlowLayout *layout = [[UICollectionViewFlowLayout alloc] init];
        _collectionView = [[UICollectionView alloc] initWithFrame:CGRectZero collectionViewLayout:layout];
        _collectionView.backgroundColor = [UIColor blackColor];
    }
    return _collectionView;
}

- (IGListAdapter *)adapter {
    if (!_adapter) {
        _adapter = [[IGListAdapter alloc] initWithUpdater:[[IGListAdapterUpdater alloc] init] viewController:self workingRangeSize:0];
    }
    return _adapter;
}

- (Pathfinder *)pathfinder {
    if (!_pathfinder) {
        _pathfinder = [[Pathfinder alloc] init];
    }
    return _pathfinder;
}

- (Weather *)currentWeather {
    if (!_currentWeather) {
        _currentWeather = [[Weather alloc] initWithTemperature:6 high:13 low:-69 date:[[NSDate alloc] init] sunrise:@"05:42" sunset:@"17:58" condition:WeatherConditionDustStorm];
    }
    return _currentWeather;
}

@end

File: Marslink_Swift5/Marslink_Swift5/ViewControllers/FeedViewController.swift
//
//  FeedViewController.swift
//  Marslink_Swift5
//
//  Created by liuruixuan on 2023/4/17.
//

import IGListDiffKit
import IGListKit
import UIKit

class FeedViewController: UIViewController {
    let loader = JournalEntryLoader()
    let collectionView: UICollectionView = {
        let view = UICollectionView(frame: .zero, collectionViewLayout: UICollectionViewFlowLayout())
        view.backgroundColor = UIColor.black
        return view
    }()

    lazy var adapter: ListAdapter = {
        ListAdapter(updater: ListAdapterUpdater(), viewController: self, workingRangeSize: 0)
    }()

    let pathfinder = Pathfinder()
    let wxScanner = WxScanner()

    override func viewDidLoad() {
        super.viewDidLoad()

        loader.loadLatest()
        view.addSubview(collectionView)
        adapter.collectionView = collectionView
        adapter.dataSource = self
        pathfinder.delegate = self
        pathfinder.connect()
    }

    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        collectionView.frame = view.bounds
    }
}

extension FeedViewController: ListAdapterDataSource {
    func objects(for listAdapter: ListAdapter) -> [ListDiffable] {
        var items: [ListDiffable] = [wxScanner.currentWeather]
        items += loader.entries as [ListDiffable]
        items += pathfinder.messages as [ListDiffable]

        return items.sorted(by: { (left: Any, right: Any) -> Bool in
            if let left = left as? DateSortable, let right = right as? DateSortable {
                return left.date > right.date
            }
            return false
        })
    }

    func listAdapter(_ listAdapter: ListAdapter, sectionControllerFor object: Any) -> ListSectionController {
        if object is Message {
            return MessageSectionController()
        } else if object is Weather {
            return WeatherSectionController()
        } else {
            return JournalSectionController()
        }
    }

    func emptyView(for listAdapter: ListAdapter) -> UIView? {
        return nil
    }
}

extension FeedViewController: PathfinderDelegate {
    func pathfinderDidUpdateMessages(pathfinder: Pathfinder) {
        adapter.performUpdates(animated: true)
    }
}
